File: app/src/main/java/jp/gr/java_conf/snake0394/loglook_android/logger/BattleLogger.java
package jp.gr.java_conf.snake0394.loglook_android.logger;

import android.os.Environment;
import android.util.Log;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStreamWriter;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.List;

import jp.gr.java_conf.snake0394.loglook_android.BattleType;
import jp.gr.java_conf.snake0394.loglook_android.bean.AbstractBattle;
import jp.gr.java_conf.snake0394.loglook_android.bean.Deck;
import jp.gr.java_conf.snake0394.loglook_android.bean.DeckManager;
import jp.gr.java_conf.snake0394.loglook_android.bean.MstShip;
import jp.gr.java_conf.snake0394.loglook_android.bean.MstShipManager;
import jp.gr.java_conf.snake0394.loglook_android.bean.MyShip;
import jp.gr.java_conf.snake0394.loglook_android.bean.MyShipManager;
import jp.gr.java_conf.snake0394.loglook_android.bean.SortieBattleresult;

/**
 * 戦闘をCSVファイルに記録します。
 */
public enum BattleLogger {
    INSTANCE;

    private boolean isFirstBattle = false;
    private int eventId;
    private int cell;
    private BattleType battleType = BattleType.NULL;
    private AbstractBattle battle;

    /**
     * api_req_map/start
     */
    public void start(int cell, int eventId) {
        isFirstBattle = true;
        this.cell = cell;
        this.eventId = eventId;
    }

    /**
     * api_req_map/next
     */
    public void next(int cell, int eventId) {
        this.cell = cell;
        this.eventId = eventId;
    }

    public void setBattle(AbstractBattle battle) {
        this.battle = battle;
    }

    public void writeLog(SortieBattleresult battleresult) {
        //SDカードのディレクトリパス
        File sdcard_path = new File(Environment.getExternalStorageDirectory().getPath() + "/泥提督支援アプリ/");

        //パス区切り用セパレータ
        String Fs = File.separator;

        //テキストファイル保存先のファイルパス
        String filePath = sdcard_path + Fs + "海戦・ドロップ報告書.csv";

        //フォルダがなければ作成
        if (!sdcard_path.exists()) {
            sdcard_path.mkdir();
        }

        try {
            StringBuffer sb = new StringBuffer();
            File file = new File(filePath);
            if (!file.exists()) {
                sb.append("日付,海域,マス,ボス,ランク,交戦形態,味方陣形,敵陣形,制空状態,味方触接機,敵触接機,敵艦隊名,ドロップ艦種,ドロップ艦娘,味方艦1,味方艦1HP,味方艦2,味方艦2HP,味方艦3,味方艦3HP,味方艦4,味方艦4HP,味方艦5,味方艦5HP,味方艦6,味方艦6HP,敵艦1,敵艦1HP,敵艦2,敵艦2HP,敵艦3,敵艦3HP,敵艦4,敵艦4HP,敵艦5,敵艦5HP,敵艦6,敵艦6HP\r\n");
            }

            BufferedWriter pw = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(filePath, true), "SJIS"));
            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
            sb.append(sdf.format(Calendar.getInstance()
                                         .getTime()) + ",");
            sb.append(battleresult.getQuestName() + ",");
            sb.append(cell + ",");
            if (isFirstBattle && eventId == 5) {
                sb.append("出撃&ボス");
            } else if (isFirstBattle) {
                sb.append("出撃");
            } else if (eventId == 5) {
                sb.append("ボス");
            } else {
                sb.append("");
            }
            sb.append(",");
            sb.append(battleresult.getRank());
            sb.append(",");
            sb.append(battle.getTactic());
            sb.append(",");
            sb.append(battle.getFormation());
            sb.append(",");
            sb.append(battle.geteFormation());
            sb.append(",");
            sb.append(battle.getSeiku());
            sb.append(",");
            sb.append(battle.getTouchPlane());
            sb.append(",");
            sb.append(battle.getEtTouchPlane());
            sb.append(",");
            sb.append(battleresult.geteFleetName());
            sb.append(",");
            sb.append(battleresult.getGetShipType());
            sb.append(",");
            sb.append(battleresult.getGetShipName());
            sb.append(",");

            Deck deck = DeckManager.INSTANCE.getDeck(battle.getDeckId());
            List<Integer> shipId = deck.getShipId();
            List<Integer> eship = battle.geteShip();
            List<Integer> nowhps = battle.getNowhps();
            List<Integer> maxhps = battle.getMaxhps();
            for (int i = 1; i < 7; i++) {
                if (shipId.get(i - 1) == -1) {
                    sb.append(",,");
                    continue;
                }
                MyShip myShip = MyShipManager.INSTANCE.getMyShip(shipId.get(i - 1));
                MstShip mstShip = MstShipManager.INSTANCE.getMstShip(myShip.getShipId());
                sb.append(mstShip.getName());
                sb.append("(Lv");
                sb.append(myShip.getLv());
                sb.append("),");
                sb.append(nowhps.get(i));
                sb.append("/");
                sb.append(maxhps.get(i));
                sb.append(",");
            }

            for (int i = 7, j = 1; i < 13; i++, j++) {
                if (eship.get(j) == -1) {
                    sb.append(",");
                } else {
                    sb.append(MstShipManager.INSTANCE.getMstShip(eship.get(j)).getName());
                    if (!MstShipManager.INSTANCE.getMstShip(eship.get(j)).getYomi().equals("") && !MstShipManager.INSTANCE.getMstShip(eship.get(j)).getYomi().equals("-")) {
                        sb.append("(");
                        sb.append(MstShipManager.INSTANCE.getMstShip(eship.get(j)).getYomi());
                        sb.append(")");
                    }
                    sb.append(",");
                    sb.append(nowhps.get(i));
                    sb.append("/");
                    sb.append((maxhps.get(i)));
                }
                if (i == 12) {
                    break;
                }
                sb.append(",");
            }
            Log.d("battlelog", sb.toString());
            isFirstBattle = false;
            sb.append("\r\n");
            pw.write(sb.toString());
            pw.flush();
            pw.close();
        } catch (Exception e) {
            e.printStackTrace();
            ErrorLogger.writeLog(e);
        }

    }
}

File: app/src/main/java/jp/gr/java_conf/snake0394/loglook_android/logger/BattleLogger.kt
package jp.gr.java_conf.snake0394.loglook_android.logger

import android.os.Environment
import android.util.Log
import com.google.gson.FieldNamingPolicy
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import io.realm.Realm
import jp.gr.java_conf.snake0394.loglook_android.BattleUtility
import jp.gr.java_conf.snake0394.loglook_android.TacticalSituation
import jp.gr.java_conf.snake0394.loglook_android.api.API
import jp.gr.java_conf.snake0394.loglook_android.api.APIListenerSpi
import jp.gr.java_conf.snake0394.loglook_android.bean.DeckManager
import jp.gr.java_conf.snake0394.loglook_android.bean.MstShip
import jp.gr.java_conf.snake0394.loglook_android.bean.MstSlotitem
import jp.gr.java_conf.snake0394.loglook_android.bean.MyShip
import jp.gr.java_conf.snake0394.loglook_android.bean.battle.*
import jp.gr.java_conf.snake0394.loglook_android.proxy.RequestMetaData
import jp.gr.java_conf.snake0394.loglook_android.proxy.ResponseMetaData
import java.io.BufferedWriter
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStreamWriter
import java.text.SimpleDateFormat
import java.util.*


/**
 * 戦闘をCSVファイルに記録します。
 */
@API("/kcsapi/api_req_map/start", "/kcsapi/api_req_map/next", "/kcsapi/api_req_sortie/battleresult", "/kcsapi/api_req_combined_battle/battleresult")
class BattleLogger : APIListenerSpi {

    private var isFirstBattle = false
    private var cell: Int = 0
    private var isBoss = false

    override fun accept(json: JsonObject, req: RequestMetaData, res: ResponseMetaData) {
        val data = json.getAsJsonObject("api_data")
        when (req.requestURI) {
            "/kcsapi/api_req_map/start" -> {
                isFirstBattle = true
                this.cell = data.get("api_no").asInt
                this.isBoss = data.get("api_event_id").asInt == 5
            }
            "/kcsapi/api_req_map/next" -> {
                this.cell = data.get("api_no").asInt
                this.isBoss = data.get("api_event_id").asInt == 5
            }
            "/kcsapi/api_req_sortie/battleresult", "/kcsapi/api_req_combined_battle/battleresult" -> {
                val result = GsonBuilder()
                        .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
                        .create()
                        .fromJson(data, SortieBattleresult::class.java)
                writeLog(result)
            }
        }
    }

    private fun writeLog(result: SortieBattleresult) {
        val sdcardPath = File(Environment.getExternalStorageDirectory().path + "/泥提督支援アプリ/")
        val filePath = sdcardPath.toString() + File.separator + "海戦・ドロップ報告書.csv"

        //フォルダがなければ作成
        if (!sdcardPath.exists()) {
            sdcardPath.mkdir()
        }

        val realm = Realm.getDefaultInstance()
        try {
            val battle = TacticalSituation.battle
            val sb = StringBuffer()
            val file = File(filePath)
            if (!file.exists()) {
                sb.append("日付,海域,マス,ボス,ランク,交戦形態,味方陣形,敵陣形,制空状態,味方触接機,敵触接機,敵艦隊名,ドロップ艦種,ドロップ艦娘,味方艦1,味方艦1HP,味方艦2,味方艦2HP,味方艦3,味方艦3HP,味方艦4,味方艦4HP,味方艦5,味方艦5HP,味方艦6,味方艦6HP,敵艦1,敵艦1HP,敵艦2,敵艦2HP,敵艦3,敵艦3HP,敵艦4,敵艦4HP,敵艦5,敵艦5HP,敵艦6,敵艦6HP\r\n")
            }

            val pw = BufferedWriter(OutputStreamWriter(FileOutputStream(filePath, true), "SJIS"))
            val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
            sb.append("${sdf.format(Calendar.getInstance().time)},${result.apiQuestName},${cell},")
            sb.append("${when {
                isFirstBattle && isBoss -> "出撃&ボス"
                isFirstBattle -> "出撃"
                isBoss -> "ボス"
                else -> ""
            }},${result.apiWinRank},")
            (battle as IFormation).run {
                sb.append("${BattleUtility.getTactic(apiFormation[2])},${BattleUtility.getFormation(apiFormation[0])},${BattleUtility.getFormation(apiFormation[1])},")
            }
            when (battle) {
                is IKouku -> {
                    (battle as IKouku).apiKouku.apiStage1!!.run {
                        sb.append("${BattleUtility.getDispSeiku(apiDispSeiku)},")
                        val touchPlane = realm.where(MstSlotitem::class.java).equalTo("id", apiTouchPlane[0]).findFirst()?.name ?: ""
                        val enemyTouchPlane = realm.where(MstSlotitem::class.java).equalTo("id", apiTouchPlane[1]).findFirst()?.name ?: ""
                        sb.append("$touchPlane,$enemyTouchPlane,")
                    }
                }
                is IMidnightBattle -> {
                    (battle as IMidnightBattle).run {
                        val touchPlane = realm.where(MstSlotitem::class.java).equalTo("id", apiTouchPlane[0]).findFirst()?.name ?: ""
                        val enemyTouchPlane = realm.where(MstSlotitem::class.java).equalTo("id", apiTouchPlane[1]).findFirst()?.name ?: ""
                        sb.append(",$touchPlane,$enemyTouchPlane,")
                    }
                }
                else -> Logger.d("BattleLogger", "illegal battle:${battle.javaClass.name}")
            }
            sb.append("${result.apiEnemyInfo.apiDeckName},${result.apiGetShip?.apiShipType ?: ""},${result.apiGetShip?.apiShipName ?: ""},")

            val deck = DeckManager.INSTANCE.getDeck(battle.apiDeckId)
            val shipId = deck.shipId
            val fNowhps = battle.apiFNowhps
            val fMaxhps = battle.apiFMaxhps
            for (i in 1..6) {
                if (shipId[i - 1] == -1) {
                    sb.append(",,")
                    continue
                }
                val myShip = realm.where(MyShip::class.java).equalTo("id", shipId[i - 1]).findFirst()
                val mstShip = realm.where(MstShip::class.java).equalTo("id", myShip.shipId).findFirst()
                sb.append("${mstShip.name}(Lv${myShip.lv}),${fNowhps[i - 1]}/${fMaxhps[i - 1]},")
            }

            val eship = battle.apiShipKe
            val eNowhps = battle.apiENowhps
            val eMaxhps = battle.apiEMaxhps
            for (i in 1..6) {
                if (i > eship.size) {
                    sb.append(",")
                } else {
                    val mstShip = realm.where(MstShip::class.java).equalTo("id", eship[i - 1]).findFirst()
                    sb.append(mstShip.name)
                    if (mstShip.yomi != "" && mstShip.yomi != "-") {
                        sb.append("(${mstShip.yomi})")
                    }
                    sb.append(",${eNowhps[i - 1]}/${eMaxhps[i - 1]}")
                }

                if (i == 6) {
                    break
                }

                sb.append(",")
            }

            sb.append(",")
            when (battle) {
                is ICombinedBattle -> {
                    val deck = DeckManager.INSTANCE.getDeck(2)
                    val shipId = deck.shipId
                    val nowhps = battle.apiNowhpsCombined
                    val maxhps = battle.apiMaxhpsCombined
                    for (i in 1..6) {
                        if (i > shipId.size) {
                            sb.append(",,")
                            continue
                        }
                        val myShip = realm.where(MyShip::class.java).equalTo("id", shipId[i - 1]).findFirst()
                        val mstShip = realm.where(MstShip::class.java).equalTo("id", myShip.shipId).findFirst()
                        sb.append("${mstShip.name}(Lv${myShip.lv}),${nowhps[i]}/${maxhps[i]},")
                    }

                    sb.append(",,,,,,,,,,,")
                }
                is IEnemyCombinedBattle -> {
                    sb.append(",,,,,,,,,,,,")

                    val nowhps = battle.apiNowhpsCombined
                    val maxhps = battle.apiMaxhpsCombined
                    val eship = battle.apiShipKeCombined
                    for (i in 1..6) {
                        if (i > eship.size) {
                            sb.append(",")
                        } else {
                            val mstShip = realm.where(MstShip::class.java).equalTo("id", eship[i - 1]).findFirst()
                            sb.append(mstShip.name)
                            if (mstShip.yomi != "" && mstShip.yomi != "-") {
                                sb.append("(${mstShip.yomi})")
                            }
                            sb.append(",${nowhps[i + 6]}/${maxhps[i + 6]}")
                        }

                        if (i == 6) {
                            break
                        }

                        sb.append(",")
                    }
                }
                is IEachCombinedBattle -> {
                    val deck = DeckManager.INSTANCE.getDeck(2)
                    val shipId = deck.shipId
                    val nowhps = battle.apiNowhpsCombined
                    val maxhps = battle.apiMaxhpsCombined
                    for (i in 1..6) {
                        if (i > shipId.size) {
                            sb.append(",,")
                            continue
                        }
                        val myShip = realm.where(MyShip::class.java).equalTo("id", shipId[i - 1]).findFirst()
                        val mstShip = realm.where(MstShip::class.java).equalTo("id", myShip.shipId).findFirst()
                        sb.append("${mstShip.name}(Lv${myShip.lv}),${nowhps[i]}/${maxhps[i]},")
                    }

                    val eship = battle.apiShipKeCombined
                    for (i in 1..6) {
                        if (i > eship.size) {
                            sb.append(",")
                        } else {
                            val mstShip = realm.where(MstShip::class.java).equalTo("id", eship[i - 1]).findFirst()
                            sb.append(mstShip.name)
                            if (mstShip.yomi != "" && mstShip.yomi != "-") {
                                sb.append("(${mstShip.yomi})")
                            }
                            sb.append(",${nowhps[i + 6]}/${maxhps[i + 6]}")
                        }

                        if (i == 6) {
                            break
                        }

                        sb.append(",")
                    }
                }
                else -> {
                    sb.append(",,,,,,,,,,,,,,,,,,,,,,,")
                }
            }

            Log.d("battlelog", sb.toString())
            sb.append("\r\n")
            pw.write(sb.toString())
            pw.flush()
            pw.close()
        } catch (e: Exception) {
            e.printStackTrace()
            ErrorLogger.writeLog(e)
        } finally {
            isFirstBattle = false
            realm.close()
        }

    }
}
